File: backend/go-api/internal/service/token_service.go
package service

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"log"
	"time"

	"saas-starter/backend/go-api/internal/domain"
	"saas-starter/backend/go-api/internal/dto"
	"saas-starter/backend/go-api/internal/platform/jwt"
	"saas-starter/backend/go-api/internal/repository"
)

type TokenService struct {
	tokens repository.TokenRepository
	jwt    *jwt.Manager
}

func NewTokenService(tokens repository.TokenRepository, jwtMgr *jwt.Manager) *TokenService {
	return &TokenService{tokens: tokens, jwt: jwtMgr}
}

// IssueTokenPair creates a new access + refresh token pair.
// A new token family is created for each login — all refresh tokens
// issued via rotation share the same family_id for theft detection.
func (s *TokenService) IssueTokenPair(ctx context.Context, userID uint, role string, rememberMe bool) (*dto.TokenPair, error) {
	familyID := mustRandomHex(16)
	return s.issueWithFamily(ctx, userID, role, rememberMe, familyID)
}

// RotateResult holds the metadata from a validated refresh token,
// allowing the caller to look up the current user role before
// issuing a new token pair.
type RotateResult struct {
	UserID   uint
	FamilyID string
}

// ValidateAndRevokeRefreshToken checks the incoming refresh token,
// revokes it, and returns the associated metadata.
// If the token was already revoked, the entire family is invalidated
// (token theft detection).
func (s *TokenService) ValidateAndRevokeRefreshToken(ctx context.Context, rawToken string) (*RotateResult, error) {
	tokenHash := domain.HashToken(rawToken)

	stored, err := s.tokens.FindRefreshTokenByHash(ctx, tokenHash)
	if err != nil {
		return nil, domain.NewError(domain.ErrTokenInvalid, "invalid refresh token")
	}

	if stored.Revoked {
		if revokeErr := s.tokens.RevokeTokenFamily(ctx, stored.FamilyID); revokeErr != nil {
			log.Printf("Failed to revoke token family %s: %v", stored.FamilyID, revokeErr)
		}
		return nil, domain.NewError(domain.ErrTokenRevoked, "token reuse detected, all sessions in this family revoked")
	}

	if time.Now().After(stored.ExpiresAt) {
		return nil, domain.NewError(domain.ErrTokenExpired, "refresh token expired")
	}

	if err := s.tokens.RevokeRefreshTokenByHash(ctx, tokenHash); err != nil {
		return nil, err
	}

	return &RotateResult{UserID: stored.UserID, FamilyID: stored.FamilyID}, nil
}

// IssueRotatedPair creates a new token pair within an existing family.
func (s *TokenService) IssueRotatedPair(ctx context.Context, userID uint, role, familyID string) (*dto.TokenPair, error) {
	return s.issueWithFamily(ctx, userID, role, false, familyID)
}

// RevokeRefreshToken revokes a single refresh token by its raw value.
func (s *TokenService) RevokeRefreshToken(ctx context.Context, rawToken string) error {
	return s.tokens.RevokeRefreshTokenByHash(ctx, domain.HashToken(rawToken))
}

// RevokeAllUserTokens revokes every active refresh token for a user.
func (s *TokenService) RevokeAllUserTokens(ctx context.Context, userID uint) error {
	return s.tokens.RevokeAllUserRefreshTokens(ctx, userID)
}

func (s *TokenService) issueWithFamily(ctx context.Context, userID uint, role string, rememberMe bool, familyID string) (*dto.TokenPair, error) {
	accessToken, err := s.jwt.Generate(userID, role, 15*time.Minute)
	if err != nil {
		return nil, err
	}

	rawRT := mustRandomHex(32)

	expiry := 24 * time.Hour
	if rememberMe {
		expiry = 7 * 24 * time.Hour
	}

	rt := &domain.RefreshToken{
		UserID:    userID,
		TokenHash: domain.HashToken(rawRT),
		FamilyID:  familyID,
		ExpiresAt: time.Now().Add(expiry),
	}
	if err := s.tokens.CreateRefreshToken(ctx, rt); err != nil {
		return nil, err
	}

	return &dto.TokenPair{
		AccessToken:  accessToken,
		RefreshToken: rawRT,
		ExpiresIn:    900,
		TokenType:    "Bearer",
	}, nil
}

func mustRandomHex(nBytes int) string {
	b := make([]byte, nBytes)
	if _, err := rand.Read(b); err != nil {
		panic("crypto/rand failed")
	}
	return hex.EncodeToString(b)
}

File: backend/python-api/app/service/token_service.py
from __future__ import annotations

import secrets
from datetime import datetime, timedelta, timezone

from app.domain.errors import ERR_TOKEN_EXPIRED, ERR_TOKEN_INVALID, ERR_TOKEN_REVOKED, DomainError
from app.domain.token import RefreshToken, hash_token
from app.dto.auth import TokenPair
from app.platform.jwt import JWTManager
from app.repository.token_repo import TokenRepository

BCRYPT_COST = 12
ACCESS_TOKEN_DURATION = timedelta(minutes=15)
REFRESH_TOKEN_DURATION = timedelta(hours=24)
REFRESH_TOKEN_REMEMBER_DURATION = timedelta(days=7)


class TokenService:
    def __init__(self, token_repo: TokenRepository, jwt_manager: JWTManager):
        self._tokens = token_repo
        self._jwt = jwt_manager

    async def issue_token_pair(self, user_id: int, role: str, remember_me: bool = False) -> TokenPair:
        family_id = secrets.token_hex(16)
        return await self._issue_with_family(user_id, role, remember_me, family_id)

    async def validate_and_revoke_refresh_token(self, raw_token: str) -> tuple[int, str]:
        """Returns (user_id, family_id). Raises on invalid/revoked/expired."""
        token_hash_val = hash_token(raw_token)
        stored = await self._tokens.find_refresh_token_by_hash(token_hash_val)

        if stored is None:
            raise DomainError(ERR_TOKEN_INVALID, "Invalid refresh token")

        if stored.revoked:
            await self._tokens.revoke_token_family(stored.family_id)
            raise DomainError(ERR_TOKEN_REVOKED, "Token reuse detected, all sessions in this family revoked")

        if datetime.now(timezone.utc) > stored.expires_at.replace(tzinfo=timezone.utc):
            raise DomainError(ERR_TOKEN_EXPIRED, "Refresh token expired")

        await self._tokens.revoke_refresh_token_by_hash(token_hash_val)
        return stored.user_id, stored.family_id

    async def issue_rotated_pair(self, user_id: int, role: str, family_id: str) -> TokenPair:
        return await self._issue_with_family(user_id, role, False, family_id)

    async def revoke_refresh_token(self, raw_token: str) -> None:
        await self._tokens.revoke_refresh_token_by_hash(hash_token(raw_token))

    async def revoke_all_user_tokens(self, user_id: int) -> None:
        await self._tokens.revoke_all_user_refresh_tokens(user_id)

    async def _issue_with_family(
        self, user_id: int, role: str, remember_me: bool, family_id: str
    ) -> TokenPair:
        access_token = self._jwt.generate(user_id, role, ACCESS_TOKEN_DURATION)
        raw_rt = secrets.token_hex(32)

        expiry = REFRESH_TOKEN_REMEMBER_DURATION if remember_me else REFRESH_TOKEN_DURATION

        rt = RefreshToken(
            user_id=user_id,
            token_hash=hash_token(raw_rt),
            family_id=family_id,
            expires_at=datetime.now(timezone.utc) + expiry,
        )
        await self._tokens.create_refresh_token(rt)

        return TokenPair(
            access_token=access_token,
            refresh_token=raw_rt,
            expires_in=900,
            token_type="Bearer",
        )
